File: 백준/Gold/1253. 좋다/좋다.java
import java.io.*;
import java.util.*;

public class Main{

    // 필요변수
    static int N;
    static int[] nums;

    // 입력
    static void input() throws Exception{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        N = Integer.parseInt(br.readLine());
        nums = new int[N];

        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < N; i++) {
            nums[i] = Integer.parseInt(st.nextToken());
        }

        // 정렬
        Arrays.sort(nums);
        br.close();
    }

    // 풀이
    static void pro() {
        int cnt = 0;

        for (int i = 0; i < N; i++) {
            int target = nums[i];
            int L = 0, R = N - 1;

            while (L < R) {
                // target이랑 같으면 L, R 이동
                if (i == L) {
                    L++;
                    continue;
                }

                if (i == R) {
                    R--;
                    continue;
                }

                int sum = nums[L] + nums[R];

                // 좋은 수 조건 달성
                if (sum == target) {
                    cnt++;
                    break;
                }

                // 달성 못한 경우
                if (sum > target) {
                    R--;
                } else {
                    L++;
                }

            }
        }

        System.out.println(cnt);
        
    }
    public static void main(String[] args) throws Exception{
        input();
        pro();
    }
}

File: 백준/Gold/1253. 좋다/좋다.cc
#include <iostream>
#include <algorithm>

using namespace std;

// 변수
int N;
int nums[2005];

void input() {
    cin >> N;

    for (int i = 0; i < N; i++) {
        cin >> nums[i];
    }

    // 정렬
    sort(nums, nums + N);
}

bool search_number(int idx) {
    int left = 0, right = N - 1;

    while (left < right) {
        // idx랑 같으면 값 이동
        if (left == idx) {
            left++;
            continue;
        }

        if (right == idx) {
            right--;
            continue;
        }

        int sum = nums[left] + nums[right];

        // 찾았다!
        if (sum == nums[idx]) {
            return true;
        }

        // 못찾으면 포인트 이동
        if (sum < nums[idx]) {
            left++;
            continue;
        }

        if (sum > nums[idx]) {
            right--;
            continue;
        }
    }

    // 찾지 못하고 while문 끝나면 없는 것
    return false;
}

void pro() {
    int cnt = 0;

    for (int i = 0; i < N; i++) {
        if (search_number(i)) {
            cnt++;
        }
    }

    // 출력
    cout << cnt;
}

int main() {
    // 최적화
    ios::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);

    // 입력
    input();

    // 실행
    pro();
}